docs/development/design/network-policy.md

```diff
@@ -0,0 +1,256 @@
+
+## Phased Implementation
+
+### Phase I - Default Security Group for Organization
+
+- The default security group for an organization will be applied to all nodes on startup.
+- A device can only have one policy group applied at any given time in phase I.
+
+### Phase II - User-owned Security Groups
+
+- Users can create their own security groups for more granular policies per node. The user owns the security group. Seems reasonable other members of an org should be able to use that security group. Similar to how an ec2 security group gets applied to a node would be the same here with a patch.
+- A device can only have one policy group applied at any given time.
+
+### Phase III - Robust Admin Organization Policies
+
+- Overarching admin policies can be overlaid on top or in lieu of individual user policies.
+- A device can potentially have multiple policy groups applied at the same time.
+
+## Default Security Group Rules
+
+The default security group rules will be an explicit allow, adopting the EC2 approach of an implicit deny. The agent will add an explicit deny at the end of the ingress or egress chain. The user only sees an implicit deny with explicit allows just as in Ec2.
+
+- Ordering will be done by the order the user installs the rules. This is possible since there are no denies. As a refernce, you can compare EC2 rules to Azure rules for not allowing deny statements vs allowing deny statements. Order begins to matter when denies are in place. This adds complexity which for our use case does not add any clear value.
+- Users can add ranges of a given field. For example both, IpRanges with a value of `100.100.0.100-100.100.0.120` is valid and a prefix such as `100.100.0.128/25` is also valid. Along with that, a single address such as `100.100.0.10`.
+- The same applies to source port and destination ports, `PortFrom:8080` coupled with `PortTo:9000` would equate to a rule of `8080-9000` being permitted. `PortFrom:0 PortTo:0` will be read as `ip permit <protocol> any`. `PortFrom:443 PortTo:443` would be equivalent to `ip permit <protocol> 4434`.
+- L3 `IpRanges` are applied based on the direction field they are located in SecurityGroups. `InboundRules` have the IP prefix applied to the `saddr` field in nftables in the input chain, while `OutboundRules` are applied to the `daddr` field in the outbound chain.
+- In regard to L4 ports, destination ports are what will be supported. For example, an ingress rule of `input tcp dport 22 counter accept` would apply port 22 to the L4 dport value in the nexodus-inbound chain meaning any host can connect to the node on port 22.
+
+## Security Group User Interface
+
+- The user can add rules via web UI, the nexctl tool or the http API. There would not be support for adding or manipulating rules via the agent.
```

Comment:
It's worth calling out here that with our current security model, there is no way to enforce that an agent can't make changes. In other words, we need device-specific tokens that are have more limited access controls. That way a compromise of a single device can't be used to make changes in the Nexodus API.

This is a problem much more broad than this specific proposal, though. If we don't have an issue, we should (I will check). Once we do have an issue, let's link it from here to explain more about how agent credentials could be prevented from making security policy changes.